### PR TITLE
feat: when handling actions, print uncaught exceptions to stderr

### LINKED ISSUE #1078: Actions: errors are captured to debug-log and are not in stderr
Setting stdout, stderr in event.set_results, always ends up with a failed action, but it never explains why it failed.

Consider the following charm:

```python
import logging

import ops

# Log messages can be retrieved using juju debug-log
logger = logging.getLogger(__name__)

VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]


class ActionTestCharm(ops.CharmBase):
    """Charm the service."""

    def __init__(self, *args):
        super().__init__(*args)
        self.framework.observe(self.on.fail_action_action, self._on_fail_action_action)
        self.framework.observe(self.on.success_action_action, self._on_success_action_action)

    def _on_fail_action_action(self, event):
        event.fail("Fail action failed")

    def _on_success_action_action(self, event):
        event.log("Success action completed successfully")
        event.set_results({"stdout":"fail me"})


if __name__ == "__main__":  # pragma: nocover
    ops.main(ActionTestCharm)  # type: ignore
```

This will always fail (against juju 3.1).

```sh
juju run action-test/leader success-action
Running operation 69 with 1 task
  - task 70 on unit-action-test-0

Waiting for task 70...
08:01:18 Success action completed successfully

ERROR the following task failed:
 - id "70" with return code 1

use 'juju show-task' to inspect the failure
```

If we run the same with bash as our invocation hook rather than python.

```bash
#!/bin/bash
action-set stdout="something something something"

echo Hello
exit 0
```

We get the following:

```sh
juju run bash-test/0 hello
Running operation 103 with 1 task
  - task 104 on unit-bash-test-0

Waiting for task 104...
08:25:21 Hello

Goodbye cruel world
ERROR cannot set reserved action key "stdout"
```

~~I'd expect it to be the same as bash, but clearly, something isn't quite right.~~

Addendum:

Setting the following in bash does reveal the same results as the framework. I guess the problem is framework doesn't expose why it failed, just that it did.

```bash
#!/bin/bash
set -eux
action-set stdout="something something something"

echo Hello
exit 0
```

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
When handling an action and an exception is not caught, output basic information about the problem to stderr, and let the user know that a full traceback (that the Charm author probably wants, rather than the user running the action) is in debug-log.

Fixes #1078

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@benhoyt three things here before this can be properly reviewed:

* [ ] I imagine there's some bikeshedding required for the argument name to `setup_root_logging` that controls this behaviour.
* [ ] We need to finalise on what the output would look like (#1078 has examples from actual crashes when doing `juju run`). My preference is on the shorter version, primarily because I think the traceback is more for the Charm author than the action running user. We could switch to the full version, and we can also tweak wording and to some extent spacing.
* [ ] What's your opinion on tests here? I don't think I can test the actual sys.excepthook in the main process. I could add a test like the main ones that have a Charm in a subprocess, get it to crash, and capture stderr. Do you think that's worthwhile? Or have a better idea about how to add an automated test for this?

**Comment by tonyandrewmeyer** [comment]
The real pebble tests fail because of the [macaroonbakery issue](https://chat.charmhub.io/charmhub/pl/w8nc98aus3fttxqqtjmmcrbace). The data charm tests fail since 2.9 it seems, although it must be some change on the other side too or we would have caught it before release - looking into that.

File: ops/log.py [hunk 1]
```diff
@@ -65,5 +67,9 @@ def except_hook(etype: typing.Type[BaseException],
         logger.error(
             "Uncaught exception while in charm code:",
             exc_info=(etype, value, tb))
+        if exc_stderr:
+            print(f"Uncaught exception while in charm code: {etype.__name__}: {value}",
```
[the remark this answers is not in the document]
Comment (tonyandrewmeyer) [reply ops/log.py:71]: Agreed, and I like that it doesn't have two colons any more.

**Comment by tonyandrewmeyer** [comment]
Notes from talking in person:

> * [ ]  I imagine there's some bikeshedding required for the argument name to `setup_root_logging` that controls this behaviour.

Existing name is fine, `setup_root_logging` probably ought to have been `_setup_root_logging` and this shouldn't impact anyone.

> * [ ]  We need to finalise on what the output would look like ([Actions: errors are captured to debug-log and are not in stderr #1078](https://github.com/canonical/operator/issues/1078) has examples from actual crashes when doing `juju run`). My preference is on the shorter version, primarily because I think the traceback is more for the Charm author than the action running user. We could switch to the full version, and we can also tweak wording and to some extent spacing.

Keep as-is other than one small suggestion.

> * [ ]  What's your opinion on tests here? I don't think I can test the actual sys.excepthook in the main process. I could add a test like the main ones that have a Charm in a subprocess, get it to crash, and capture stderr. Do you think that's worthwhile? Or have a better idea about how to add an automated test for this?

Add a `test_main` test that has a failing action, check for the core pieces rather than character-exact matching.

File: test/test_main.py [hunk 2]
```diff
@@ -1128,10 +1128,22 @@ def _call_event(self, rel_path: Path, env: typing.Dict[str, str]):
...
+        self._prepare_actions()
+        self.stderr = tempfile.TemporaryFile()
+        self.addCleanup(self.stderr.close)
+        fake_script(typing.cast(unittest.TestCase, self), 'action-get', "echo '{}'")
+        with self.assertRaises(subprocess.CalledProcessError):
+            self._simulate_event(EventSpec(
+                ops.ActionEvent, 'keyerror_action',
+                env_var='JUJU_ACTION_NAME'))
+        self.stderr.seek(0)
+        self.assertIn(b'KeyError', self.stderr.read())
```
Comment (benhoyt) [inline test/test_main.py:1145]: I also think we should test that the exception value (string) is included in the output (just not the "Uncaught" and "Use juju debug-log" bits). Perhaps:

```suggestion
        stderr = self.stderr.read()
        self.assertIn(b'KeyError', stderr)
        self.assertIn(b"'foo' not found in 'bar'", stderr)
```

**Review by benhoyt — CHANGES_REQUESTED** [review]
Looks great to me. One test change suggested.

**Review by benhoyt — APPROVED** [review]

**Pull request merged**